src-java/swmanager-topology/swmanager-storm-topology/src/main/java/org/openkilda/wfm/topology/switchmanager/service/impl/ValidationServiceImpl.java

```diff
@@ -137,86 +163,259 @@ public ValidateRulesResult validateRules(SwitchId switchId, List<FlowSpeakerData
+                                                 boolean excludeFlowInfo) {
+        log.debug("Validating meters on a switch {}", switchId);
+
+        Switch sw = switchRepository.findById(switchId)
+                .orElseThrow(() -> new SwitchNotFoundException(switchId));
+        boolean isESwitch = Switch.isNoviflowESwitch(sw.getOfDescriptionManufacturer(), sw.getOfDescriptionHardware());
+
+        List<MeterInfoEntryV2> actualMeters =  convertMeters(switchId, presentMeters,
+                excludeFlowInfo);
+        List<MeterInfoEntryV2> expectedMeters = convertMeters(switchId, expectedMeterSpeakerData,
+                excludeFlowInfo);
+
+        List<MeterInfoEntryV2> missingMeters = new ArrayList<>();
+        List<MeterInfoEntryV2> properMeters = new ArrayList<>();
+        List<MeterInfoEntryV2> excessMeters = new ArrayList<>();
+        List<MisconfiguredInfo<MeterInfoEntryV2>> misconfiguredMeters = new ArrayList<>();
+
+        processMeterValidation(isESwitch, actualMeters, expectedMeters, missingMeters, properMeters, excessMeters,
+                misconfiguredMeters);
+
+        if (!missingMeters.isEmpty() && log.isErrorEnabled()) {
+            log.warn("On switch {} the following meters are missed: {}", switchId,
+                    metersIntoLogRepresentation(missingMeters));
+        }
+
+        if (!excessMeters.isEmpty() && log.isWarnEnabled()) {
+            log.warn("On switch {} the following meters are excessive: {}", switchId,
+                    metersIntoLogRepresentation(excessMeters));
+        }
+
+        if (!misconfiguredMeters.isEmpty() && log.isWarnEnabled()) {
+            for (MisconfiguredInfo<MeterInfoEntryV2> meter : misconfiguredMeters) {
+                log.warn("On switch {} meter {} is misconfigured: {}", switchId, meter.getId(),
+                        getMisconfiguredMeterDifferenceAsString(meter.getExpected(), meter.getDiscrepancies()));
+            }
+        }
+
+        return new ValidateMetersResultV2(
+                missingMeters.isEmpty() && misconfiguredMeters.isEmpty() && excessMeters.isEmpty(),
+                missingMeters, properMeters, excessMeters, misconfiguredMeters);
+    }
+
+    private void processGroupsValidation(Set<GroupInfoEntryV2> presentGroups, Set<GroupInfoEntryV2> expectedGroups,
+                                         Set<GroupInfoEntryV2> missingGroups, Set<GroupInfoEntryV2> properGroups,
+                                         Set<GroupInfoEntryV2> excessGroups,
+                                         Set<MisconfiguredInfo<GroupInfoEntryV2>> misconfiguredGroups) {
+        Set<Integer> presentGroupsIds = presentGroups.stream()
+                .map(GroupInfoEntryV2::getGroupId)
+                .collect(Collectors.toSet());
+
+        expectedGroups.stream()
+                .filter(entry -> !presentGroupsIds.contains(entry.getGroupId()))
+                .forEach(missingGroups::add);
+
+        properGroups.retainAll(presentGroups);
```

Comment:
Group validation looks incomplete - it check consider groupId while other kind of validations make attempts to consider all object fields (as far as I see).

PS Now it is possible to include group into "proper" and into "misconfigured" lists simultaneously.